sympy/physics/units/dimensions.py

```diff
@@ -150,20 +151,17 @@ def _register_as_base_dim(self):
         dimsys_default._args = (dimsys_default.args[:2] + (Dict(d),))
 
     def __add__(self, other):
-        """Define the addition for Dimension.
-
-        Addition of dimension has a sense only if the second object is
-        the same dimension (we don't add length to time).
-        """
-
-        if not isinstance(other, Dimension):
-            raise TypeError("Only dimension can be added; '%s' is not valid"
-                            % type(other))
-        elif isinstance(other, Dimension) and self != other:
-            raise ValueError("Only dimension which are equal can be added; "
-                             "'%s' and '%s' are different" % (self, other))
+        from sympy.physics.units.quantities import Quantity
+        if isinstance(other, Basic) and other.has(Quantity):
+             other = Dimension(Quantity.get_dimensional_expr(other))
+        if self == other:
+            return self
+        elif isinstance(other, Dimension):
```

Comment:
Does this mean that the exception is only raised if both, self and other are dimensions, i.e. not for e.g. `1 + length`?

Reply:
yes, for the same reason as for allowing `1 + foot` (see answer below for motivation)